Correctly resolve custom host and port with server resources

mattpap opened the pull request:
The solution works and allows to correctly resolve host/port with server resources. ~~The only issue is with handling all the incidental complexity around resources~~.

fixes #12400

mattpap committed:
Correctly resolve custom host and port with server resources

bryevdv commented on src/bokeh/embed/bundle.py:
```diff
@@ -159,12 +159,12 @@ def bundle_for_objs_and_resources(objs: Sequence[Model | Document] | None,
+
+    if isinstance(resources, BaseResources):
         js_resources = css_resources = resources
     elif isinstance(resources, tuple) and len(resources) == 2 and all(r is None or isinstance(r, BaseResources) for r in resources):
```
We could do a git search to try to see if anyone anywhere is utilizing this complexity. I really don't expect anyone is even using `bundle_for_objs_and_resources` since we don't actually include it in the documentation at all. I would very much support drastically streamlining these codepaths.

mattpap committed:
Use pathlib.Path instead of a comment

mattpap committed:
Fix argument types in _bundle_extensions()

mattpap committed:
Remove support for BaseResources

mattpap committed:
Add a release note

bryevdv commented on docs/bokeh/source/docs/releases/3.2.0.rst:
```diff
@@ -6,3 +6,9 @@
+----------------
+
+* Removed support for ``BaseResources``, ``JSResources`` and ``CSSResources``
+  from resources, embedding and bundling APIs.
```
Reminder this info now goes in the Wiki so that it can be easily edited post-hoc if need be. 

https://github.com/bokeh/bokeh/wiki/Migration-Guides

I'm not sure how easy it would be to update the Wiki in a way that leave a commit, etc that could be linked in a PR to establish the paper trail, but that would be nice if so

mattpap replied:
I updated the wiki.

bryevdv replied:
I would also remove the info here, but I guess we can do that at release time

mattpap replied:
My understanding is that this file contains notification of changes, whereas the wiki contains both this information and  migration path(s), workarounds and/or other potentially useful information to act upon those changes.

bryevdv replied:
That's not my recollection and does not reflect the current state of the releases page:

https://docs.bokeh.org/en/latest/docs/releases.html

There we now generally present a curated list of *high-level* descriptions of changes, as links to specific PRs or issues. So, e.g. here a bullet for:

> * Correctly resolve custom host and port with server resources [#13041](https://github.com/bokeh/bokeh/pull/13041#top)

would be appropriate

mattpap replied:
I updated the changelog.

bryevdv commented on src/bokeh/embed/bundle.py:
```diff
@@ -44,10 +45,9 @@
 from ..model import Model
-from ..resources import BaseResources, Resources
+from ..resources import Resources
 from ..settings import settings
```
Is this import unused? I searched the entire file changes and I didn't see `settings` referenced anywhere

mattpap replied:
`settings.dev` is used in `_bundle_extensions()`.

bryevdv commented on src/bokeh/embed/bundle.py:
```diff
@@ -197,28 +179,25 @@ def bundle_for_objs_and_resources(objs: Sequence[Model | Document] | None,
-        extensions = _bundle_extensions(all_objs if objs else None, js_resources)
-        mode = js_resources.mode if resources is not None else "inline"
+    if resources is not None:
+        resources = deepcopy(resources)
```
I know this is how it was previously, but I guess a `clone` method with tests would be a little more confidence inspiring. It's just easy to imagine we unintentionally introduce something that is not handled well by default `deepcopy`

mattpap replied:
I implemented `Resources.clone()` and refactored this code.

bryevdv commented on src/bokeh/embed/bundle.py:
```diff
@@ -145,46 +145,28 @@ def add(self, artifact: Artifact) -> None:
 
     '''
-    # Any env vars will overide a local default passed in
-    resources = settings.resources(default=resources)
```
Just double checking, this won't prevent e.g. 
```
BOKEH_RESOURCES=inline python foo.py
```
correct? I use that constantly and have advised others on it many many times. I'm guessing the codepaths in `bokeh/resources.py` take care of this, but it would be nice to confirm (and it which case decoupling the settings from here is terrific)

mattpap replied:
I use this together with `BOKEH_DEV` exclusively for development. I tested all combinations and it all works as expected.

bryevdv reviewed (changes requested):
Looks great, only a few comments

mattpap committed:
Add support for Resources.clone()

mattpap committed:
Include settings.dev in Resources.dev

Allows to specify:

BOKEH_DEV=true BOKEH_RESOURCES=server python ...

instead of:

BOKEH_DEV=true BOKEH_RESOURCES=server-dev python ...

mattpap commented:
Tests are failing due to #13056.
